### ISSUE TITLE
re/pat_advanced.t test 396 fails under -Dusethreads -Duse64bitall (solaris)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#14640]
Migrated from [rt.perl.org#124348](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=124348) (status was 'resolved')

Searchable as RT124348$

**Comment by p5pRT** [comment]
### From @tonycoz

#### Created by @tonycoz



This requires both \-Dusethread and \-Duse64bitall and does not occur
with \-DDEBUGGING\.

This was bisected to&#8203;:

commit 64935bc6975bb01af403817752e88d6540c8711d
Author&#8203;: Karl Williamson \<khw@&#8203;cpan\.org>
Date&#8203;:   Tue Feb 17 15&#8203;:03&#8203;:32 2015 \-0700

&nbsp;    Add qr/\\b\{gcb\}/
&nbsp;    
&nbsp;    A function implements seeing if the space between any two characters is
&nbsp;    a grapheme cluster break\.  Afer I wrote this\, I realized that an array
&nbsp;    lookup might be a better implementation\, but the deadline for v5\.22 was
&nbsp;    too close to change it\.  I did see that my gcc optimized it down to
&nbsp;    an array lookup\.
&nbsp;    
&nbsp;    This makes the implementation of \\X go from being complicated to
&nbsp;    trivial\.

Setting \-Doptimize=\-g also prevents it from occuring\.

Separating out the test into a separate file\, like&#8203;:

\#\!perl \-l
package Str;
use overload q /""/ => sub \{$\{$\_ \[0\]\};\};
sub new \{my \($c\, $v\) = @&#8203;\_; bless \\$v\, $c;\}

package main;
$\_ = Str \-> new \("a\\x\{100\}/\\x\{100\}b"\);
my @&#8203;temp = /\\b\(\.\)\\x\{100\}/g;
print join \("&#8203;:"\, @&#8203;temp\); \#\, "a&#8203;:/"\, "re\_intuit\_start and PL\_bostr";

produces&#8203;:

/

as the output\.

The correct result \("a&#8203;:/"\) is produced if a non\-overloaded value is
used\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl 5.21.11:

Configured by tony at Mon Apr 20 16:42:53 EST 2015.

Summary of my perl5 (revision 5 version 21 subversion 11) configuration:
  Commit id: 6e8a73f25951830af805a38e0f612768bfcda4a5
  Platform:
    osname=solaris, osvers=2.11, archname=i86pc-solaris-thread-multi-64
    uname='sunos nereid 5.11 11.1 i86pc i386 i86pc '
    config_args='-des -Dusedevel -Dusethreads -Duse64bitall'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=define, usemultiplicity=define
    use64bitint=define, use64bitall=define, uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='/opt/solarisstudio12.3/bin/cc', ccflags ='-D_REENTRANT -m64 -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64 -DPERL_USE_SAFE_PUTENV',
    optimize='-O',
    cppflags='-D_REENTRANT -m64'
    ccversion='Sun C 5.12 SunOS_i386 2011/11/16', gccversion='', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678, doublekind=3
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16, longdblkind=3
    ivtype='long', ivsize=8, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, prototype=define
  Linker and Libraries:
    ld='/opt/solarisstudio12.3/bin/cc', ldflags =' -m64 -L/opt/solarisstudio12.3/prod/lib/amd64 -L/lib/64 -L/usr/gnu/lib '
    libpth=/opt/solarisstudio12.3/prod/lib/amd64 /lib/64 /usr/gnu/lib /usr/lib /usr/ccs/lib
    libs=-lpthread -lsocket -lnsl -lgdbm -ldb -ldl -lm -lc
    perllibs=-lpthread -lsocket -lnsl -ldl -lm -lc
    libc=/lib/libc.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-KPIC', lddlflags=' -G -m64 -L/opt/solarisstudio12.3/prod/lib/amd64 -L/lib/64 -L/usr/gnu/lib'



@INC for perl 5.21.11:
    lib
    /opt/lib/perl5/site_perl/5.21.11/i86pc-solaris-thread-multi-64
    /opt/lib/perl5/site_perl/5.21.11
    /opt/lib/perl5/5.21.11/i86pc-solaris-thread-multi-64
    /opt/lib/perl5/5.21.11
    .


Environment for perl 5.21.11:
    HOME=/home/tony
    LANG=C
    LANGUAGE (unset)
    LC_ALL=
    LC_COLLATE=
    LC_CTYPE=
    LC_MESSAGES=
    LC_MONETARY=
    LC_NUMERIC=
    LC_TIME=
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/opt/solarisstudio12.3/bin/:/usr/bin:/usr/sbin
    PERL_BADLANG (unset)
    SHELL=/usr/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @khwilliamson

On 04/20/2015 01&#8203;:21 AM\, Tony Cook \(via RT\) wrote&#8203;:
> \# New Ticket Created by  Tony Cook
> \# Please include the string&#8203;:  \[perl \#124348\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=124348 >
>
>
> This is a bug report for perl from tony@&#8203;develop\-help\.com\,
> generated with the help of perlbug 1\.40 running under perl 5\.21\.11\.
>
>
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please describe your issue here\]
>
> This requires both \-Dusethread and \-Duse64bitall and does not occur
> with \-DDEBUGGING\.
>
> This was bisected to&#8203;:
>
> commit 64935bc6975bb01af403817752e88d6540c8711d
> Author&#8203;: Karl Williamson \<khw@&#8203;cpan\.org>
> Date&#8203;:   Tue Feb 17 15&#8203;:03&#8203;:32 2015 \-0700
>
>      Add qr/\\b\{gcb\}/
>
>      A function implements seeing if the space between any two characters is
>      a grapheme cluster break\.  Afer I wrote this\, I realized that an array
>      lookup might be a better implementation\, but the deadline for v5\.22 was
>      too close to change it\.  I did see that my gcc optimized it down to
>      an array lookup\.
>
>      This makes the implementation of \\X go from being complicated to
>      trivial\.
>
> Setting \-Doptimize=\-g also prevents it from occuring\.
>
> Separating out the test into a separate file\, like&#8203;:
>
> \#\!perl \-l
> package Str;
> use overload q /""/ => sub \{$\{$\_ \[0\]\};\};
> sub new \{my \($c\, $v\) = @&#8203;\_; bless \\$v\, $c;\}
>
> package main;
> $\_ = Str \-> new \("a\\x\{100\}/\\x\{100\}b"\);
> my @&#8203;temp = /\\b\(\.\)\\x\{100\}/g;
> print join \("&#8203;:"\, @&#8203;temp\); \#\, "a&#8203;:/"\, "re\_intuit\_start and PL\_bostr";
>
> produces&#8203;:
>
> /
>
> as the output\.
>
> The correct result \("a&#8203;:/"\) is produced if a non\-overloaded value is
> used\.
>
>

Is this only on solaris?  Is it a 5\.22 blocker in your opinion?

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @tonycoz

On Mon Apr 20 10&#8203;:11&#8203;:55 2015\, public@&#8203;khwilliamson\.com wrote&#8203;:
> On 04/20/2015 01&#8203;:21 AM\, Tony Cook \(via RT\) wrote&#8203;:
> > The correct result \("a&#8203;:/"\) is produced if a non\-overloaded value is
> > used\.
> 
> Is this only on solaris?  Is it a 5\.22 blocker in your opinion?

I think so\, both use64bitall and usethreads are common\.

The smoke has been consistently failing since the commit\.

I'll get you access to the VM where this is happening\, if you like\.

Tony

**Comment by p5pRT** [comment]
### From @khwilliamson

On 04/20/2015 04&#8203;:12 PM\, Tony Cook via RT wrote&#8203;:
> On Mon Apr 20 10&#8203;:11&#8203;:55 2015\, public@&#8203;khwilliamson\.com wrote&#8203;:
>> On 04/20/2015 01&#8203;:21 AM\, Tony Cook \(via RT\) wrote&#8203;:
>>> The correct result \("a&#8203;:/"\) is produced if a non\-overloaded value is
>>> used\.
>>
>> Is this only on solaris?  Is it a 5\.22 blocker in your opinion?
>
> I think so\, both use64bitall and usethreads are common\.
>
> The smoke has been consistently failing since the commit\.
>
> I'll get you access to the VM where this is happening\, if you like\.
>
> Tony
>
> \-\-\-
> via perlbug&#8203;:  queue&#8203;: perl5 status&#8203;: open
> https://rt-archive.perl.org/perl5/Ticket/Display.html?id=124348
>

I now have access\, and can reproduce the problem\.  Interestingly it goes 
away if 'use re qw\(Debug COMPILE\)' is added\.

But I can't find a gdb on the system to use to trace the execution\.

**Comment by p5pRT** [comment]
### From @iabyn

On Tue\, Apr 21\, 2015 at 12&#8203;:30&#8203;:24PM \-0600\, Karl Williamson wrote&#8203;:
> But I can't find a gdb on the system to use to trace the execution\.

IIRC \(and I probably don't\) the Solaris equivalent of gdb is dbx or
something like that\.

\-\- 
The optimist believes that he lives in the best of all possible worlds\.
As does the pessimist\.

**Comment by p5pRT** [comment]
### From @tonycoz

On Wed\, Apr 22\, 2015 at 11&#8203;:04&#8203;:43AM \+0100\, Dave Mitchell wrote&#8203;:
> On Tue\, Apr 21\, 2015 at 12&#8203;:30&#8203;:24PM \-0600\, Karl Williamson wrote&#8203;:
> > But I can't find a gdb on the system to use to trace the execution\.
> 
> IIRC \(and I probably don't\) the Solaris equivalent of gdb is dbx or
> something like that\.

adb\, which on Solaris 11 is a link to mdb\.

But I've installed gdb\.

Tony

**Comment by p5pRT** [comment]
### From @tux

On Wed\, 22 Apr 2015 20&#8203;:54&#8203;:27 \+1000\, Tony Cook \<tony@&#8203;develop\-help\.com>
wrote&#8203;:

> On Wed\, Apr 22\, 2015 at 11&#8203;:04&#8203;:43AM \+0100\, Dave Mitchell wrote&#8203;:
> > On Tue\, Apr 21\, 2015 at 12&#8203;:30&#8203;:24PM \-0600\, Karl Williamson wrote&#8203;:
> > > But I can't find a gdb on the system to use to trace the execution\.
> > 
> > IIRC \(and I probably don't\) the Solaris equivalent of gdb is dbx or
> > something like that\.
> 
> adb\, which on Solaris 11 is a link to mdb\.

adb \(absolute debugger or advanced debugger\) is more of an old\-style
debugger  present on most old\-style OS's and is has a completely
different interface but supports both COFF and ELF\. The advanced
debugger adb is the standard UNIX debugger found on Solaris 1 & 2\,
HP\-UX and SCO\. It is the successor of a debugger called db\.

dbx is a source\-level debugger found primarily on Solaris\, AIX\, IRIX\,
Tru64 UNIX\, Linux and BSD operating systems\. dbx was originally
developed at University of California\, Berkeley\, and subsequently made
its way to various vendors who had licensed BSD\. dbx supports DWARF\.

aix $ dbx \-\-help
dbx&#8203;: fatal error&#8203;: DBX Startup Options&#8203;:

dbx \[\-a ProcessID\] \[\-B DebugFile\] \[\-c CommandFile\] \[\-I Directory\]
\[ \-E DebugEnvironment \] \[\-p \[OldPath=NewPath&#8203;:\.\.\. | File\]\] \[\-k\]
\[\-v\] \[\-u\] \[\-x\] \[\-F\] \[\-L\] \[\-r\] \[\-C CoreFile | ObjectFile \[CoreFile\]\]

&nbsp;        \-a ProcessID        Attach to specified process
&nbsp;        \-c CommandFile      Run dbx subcommands in specified file first
&nbsp;        \-I Directory        Include Directory in list of directories
&nbsp;                            searched for source files
&nbsp;        \-C CoreFile         Allow to analyze core dump without ObjectFile
&nbsp;        \-p OldPath=NewPath  Substitute library path for core examination
&nbsp;                            or when attaching to a process
&nbsp;           File             Read library path substitutions for core
&nbsp;                            examination or when attaching to a process
&nbsp;                            from File
&nbsp;        \-E DebugEnvironment Specifies the environment variable for the
&nbsp;                            debug program
&nbsp;        \-B DebugFile        Specify an alternate debug file on startup
&nbsp;        \-v                  Relax core file validity checking
&nbsp;        \-k                  Map memory addresses
&nbsp;        \-u                  Prepend file name symbols with an '@&#8203;'
&nbsp;        \-x                  Strip postfix '\_' from FORTRAN symbols
&nbsp;        \-F                  Read all symbols at start\-up time
&nbsp;        \-L                  Keep linkage symbols
&nbsp;        \-r                  Run object file immediately

aix $ adb \-\-help
1281\-061&#8203;: Usage&#8203;: adb \[ \-w \] \[ \-Idir \] \[ \-k \] \[ \-pPROMPT \] \[ a\.out \[ core \]\]

hpux $ adb \-\-help
cannot open \`\-\-help'
hpux $ man adb 
 NAME
&nbsp;      adb \- absolute debugger

 SYNOPSIS
&nbsp;      adb \[\-w\] \[\-Idir\] \[\-k\] \[\-m\] \[\-Ppid\] objfil \[corfil\]

HP bundles a set of debuggers with an umbrella GUI wdb \(Wildebeast
Debugger\)\.

$ wdb \-version
HP WDB\-GUI \- HP Debugger Graphical User Interface \- Version 1\.1
$ wdb \-?
usage&#8203;: wdb \[gdb options\] \[Xt options\] \[executable\-file \[core\-file or process\-id\]\]

$ dde \-v
dde\, version 4\.04\.Z2

$ xdb \-v
HP92453\-02A\.10\.10/A\.10\.20 HP\-UX SYMBOLIC DEBUGGER \(XDB\) $Revision&#8203;: 75\.08 $

 NAME
&nbsp;      xdb \- C\, FORTRAN\, Pascal\, and C\+\+ Symbolic Debugger

 SYNOPSIS
&nbsp;      xdb \[\-d dir\] \[\-r file\] \[\-R file\] \[\-p file\] \[\-P process\_ID\] \[\-L\] \[\-l
&nbsp;      library\]
&nbsp;          \[\-i file\] \[\-o file\] \[\-e file\] \[\-S num\] \[\-s\] \[objectfile \[corefile\]\]

$ xdb \-\-help
Usage&#8203;:  xdb \[options \.\.\.\] \[objfile \[corefile\]\]

&nbsp;        Options&#8203;:
&nbsp;                \-d dir          specify alternate source directory
&nbsp;                \-r file         record debugger commands to file
&nbsp;                \-p file         playback debugger commands from file
&nbsp;                \-R file         restore\-state file \(from 'ss' command\)
&nbsp;                \-P pid          adopt a running process
&nbsp;                \-i path         redirect child stdin to file or device
&nbsp;                \-o path         redirect child stdout to file or device
&nbsp;                \-e path         redirect child stderr to file or device
&nbsp;                \-s              enable debugging of shared libraries
&nbsp;                \-l lib          pre\-load data from named shared library
&nbsp;                \-S num          increase string\-cache size
&nbsp;                \-L              line mode \(default is screen mode\)
&nbsp;                \-v              print version and exit

$ /opt/langtools/bin/pxdb \-\-help
/opt/langtools/bin/pxdb&#8203;: illegal option \-\- \-
Usage&#8203;:    pxdb \[\-m\]\[\-v\] progfile
&nbsp;          pxdb \[\-x dbinfofile \[\-t statusfile\] \[\-f listfile\]\] objfile \.\.\.
&nbsp;          pxdb \[\-m\] \[\-s \{on|off|enable|disable|status\}\] progfile

and finally GNU gdb

$ /opt/langtools/bin/gdb \-\-version
HP gdb 1\.1
Copyright 1986 \- 1999 Free Software Foundation\, Inc\.
Hewlett\-Packard Wildebeest 1\.1 \(based on GDB 4\.17\-hpwdb\-980821\)


> But I've installed gdb\.
> 
> Tony

\-\- 
H\.Merijn Brand  http&#8203;://tux\.nl   Perl Monger  http&#8203;://amsterdam\.pm\.org/
using perl5\.00307 \.\. 5\.21   porting perl5 on HP\-UX\, AIX\, and openSUSE
http&#8203;://mirrors\.develooper\.com/hpux/        http&#8203;://www\.test\-smoke\.org/
http&#8203;://qa\.perl\.org   http&#8203;://www\.goldmark\.org/jeff/stupid\-disclaimers/

**Comment by p5pRT** [comment]
### From @khwilliamson

On 04/20/2015 04&#8203;:12 PM\, Tony Cook via RT wrote&#8203;:
> On Mon Apr 20 10&#8203;:11&#8203;:55 2015\, public@&#8203;khwilliamson\.com wrote&#8203;:
>> On 04/20/2015 01&#8203;:21 AM\, Tony Cook \(via RT\) wrote&#8203;:
>>> The correct result \("a&#8203;:/"\) is produced if a non\-overloaded value is
>>> used\.
>>
>> Is this only on solaris?  Is it a 5\.22 blocker in your opinion?
>
> I think so\, both use64bitall and usethreads are common\.
>
> The smoke has been consistently failing since the commit\.
>
> I'll get you access to the VM where this is happening\, if you like\.
>
> Tony
>
> \-\-\-
> via perlbug&#8203;:  queue&#8203;: perl5 status&#8203;: open
> https://rt-archive.perl.org/perl5/Ticket/Display.html?id=124348
>

The latest info about this is&#8203;:

Turning on compile and/or execute re debugging causes it to not fail\.

Compiling with \-O1 doesn't fail\.

perl won't fully compile with \-O0\, failing at the last moment in 
mktables \(while writing out the list of files that it changed\)\.

If I don't clean the workspace from an \-O1\, but re\-run Configure with 
instead \-O0\, then 'make test' \(so that mktables doesn't run\)\, the 
project compiles\, but with many many test failures\.

**Comment by p5pRT** [comment]
### From pdb_ml@yahoo.com.au

On Wed\, Apr 22\, 2015 4&#8203;:24&#8203;:52 PM\, public@&#8203;khwilliamson\.com wrote&#8203;:
> On 04/20/2015 04&#8203;:12 PM\, Tony Cook via RT wrote&#8203;:
> > On Mon Apr 20 10&#8203;:11&#8203;:55 2015\, public@&#8203;khwilliamson\.com wrote&#8203;:
> >> On 04/20/2015 01&#8203;:21 AM\, Tony Cook \(via RT\) wrote&#8203;:
> >>> The correct result \("a&#8203;:/"\) is produced if a non\-overloaded value is
> >>> used\.
> >>
> >> Is this only on solaris?  Is it a 5\.22 blocker in your opinion?
> >
> > I think so\, both use64bitall and usethreads are common\.
> >
> > The smoke has been consistently failing since the commit\.
> >
> > I'll get you access to the VM where this is happening\, if you like\.
> >
> > Tony
> >
> > \-\-\-
> > via perlbug&#8203;:  queue&#8203;: perl5 status&#8203;: open
> > https://rt-archive.perl.org/perl5/Ticket/Display.html?id=124348
> >
> 
> The latest info about this is&#8203;:
> 
> Turning on compile and/or execute re debugging causes it to not fail\.
> 
> Compiling with \-O1 doesn't fail\.
> 
> perl won't fully compile with \-O0\, failing at the last moment in 
> mktables \(while writing out the list of files that it changed\)\.
> 
> If I don't clean the workspace from an \-O1\, but re\-run Configure with 
> instead \-O0\, then 'make test' \(so that mktables doesn't run\)\, the 
> project compiles\, but with many many test failures\.
> 

Just some additional information\,

TL;DR&#8203;: this bug does not seem to affect perl\-5\.21\.11 built with GCC\.

Yesterday I built 12 variants of perl 5\.21\.11 using GCC 4\.8\.4 and GCC 4\.9\.2\, on X86 32\-bit and 64\-bit Compile Servers\, running Sun Solaris 10 Update 8\, Oracle Solaris 10 Update 11 and Oracle Solaris 11\.2\. No failures seen in any "gmake test" runs\. These GCC were locally built not vendor provided\. The 32\-bit builds trigger a much larger set of warnings than the 64\-bit builds but no compilation aborts\.

64\-bit systems using \-DDuse64bitall
32\-bit systems using \-Duse64bitint
All systems using \-Dusethreads \-Dusedtrace

Other options to change installation paths\, use bison\, and be explicit about a few Solaris default like \-Duselargefiles\.
Before starting Configure\, removed all instances of \-G from hints/solaris\_2\.h \(overkill I know\) and used \-Dlddlflags='\-shared'

Regards\,
Peter

PS&#8203;: VM Host names show the combinations explicitly\.

sol10u08\-gcc48\-x86\-32
sol10u08\-gcc48\-x86\-64
sol10u08\-gcc49\-x86\-32
sol10u08\-gcc49\-x86\-64
sol10u11\-gcc48\-x86\-32
sol10u11\-gcc48\-x86\-64
sol10u11\-gcc49\-x86\-32
sol10u11\-gcc49\-x86\-64
sol11u02\-gcc48\-x86\-32
sol11u02\-gcc48\-x86\-64
sol11u02\-gcc49\-x86\-32
sol11u02\-gcc49\-x86\-64

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Wed\, Apr 22\, 2015 4&#8203;:24&#8203;:52 PM\, public@&#8203;khwilliamson\.com wrote&#8203;:
> On 04/20/2015 04&#8203;:12 PM\, Tony Cook via RT wrote&#8203;:
> > On Mon Apr 20 10&#8203;:11&#8203;:55 2015\, public@&#8203;khwilliamson\.com wrote&#8203;:
> >> On 04/20/2015 01&#8203;:21 AM\, Tony Cook \(via RT\) wrote&#8203;:
> >>> The correct result \("a&#8203;:/"\) is produced if a non\-overloaded value is
> >>> used\.
> >>
> >> Is this only on solaris?  Is it a 5\.22 blocker in your opinion?
> >
> > I think so\, both use64bitall and usethreads are common\.
> >
> > The smoke has been consistently failing since the commit\.
> >
> > I'll get you access to the VM where this is happening\, if you like\.
> >
> > Tony
> >
> > \-\-\-
> > via perlbug&#8203;:  queue&#8203;: perl5 status&#8203;: open
> > https://rt-archive.perl.org/perl5/Ticket/Display.html?id=124348
> >
> 
> The latest info about this is&#8203;:
> 
> Turning on compile and/or execute re debugging causes it to not fail\.
> 
> Compiling with \-O1 doesn't fail\.
> 
> perl won't fully compile with \-O0\, failing at the last moment in 
> mktables \(while writing out the list of files that it changed\)\.
> 
> If I don't clean the workspace from an \-O1\, but re\-run Configure with 
> instead \-O0\, then 'make test' \(so that mktables doesn't run\)\, the 
> project compiles\, but with many many test failures\.
> 

Just some additional information\,

TL;DR&#8203;: this bug does not seem to affect perl\-5\.21\.11 built with GCC\.

Yesterday I built 12 variants of perl 5\.21\.11 using GCC 4\.8\.4 and GCC 4\.9\.2\, on X86 32\-bit and 64\-bit Compile Servers\, running Sun Solaris 10 Update 8\, Oracle Solaris 10 Update 11 and Oracle Solaris 11\.2\. No failures seen in any "gmake test" runs\. These GCC were locally built not vendor provided\. The 32\-bit builds trigger a much larger set of warnings than the 64\-bit builds but no compilation aborts\.

64\-bit systems using \-DDuse64bitall
32\-bit systems using \-Duse64bitint
All systems using \-Dusethreads \-Dusedtrace

Other options to change installation paths\, use bison\, and be explicit about a few Solaris default like \-Duselargefiles\.
Before starting Configure\, removed all instances of \-G from hints/solaris\_2\.h \(overkill I know\) and used \-Dlddlflags='\-shared'

Regards\,
Peter

PS&#8203;: VM Host names show the combinations explicitly\.

sol10u08\-gcc48\-x86\-32
sol10u08\-gcc48\-x86\-64
sol10u08\-gcc49\-x86\-32
sol10u08\-gcc49\-x86\-64
sol10u11\-gcc48\-x86\-32
sol10u11\-gcc48\-x86\-64
sol10u11\-gcc49\-x86\-32
sol10u11\-gcc49\-x86\-64
sol11u02\-gcc48\-x86\-32
sol11u02\-gcc48\-x86\-64
sol11u02\-gcc49\-x86\-32
sol11u02\-gcc49\-x86\-64

**Comment by p5pRT** [comment]
### From @tux

On Thu\, 23 Apr 2015 18&#8203;:02&#8203;:37 \-0700\, "Peter Bray via RT"
\<perlbug\-comment@&#8203;perl\.org> wrote&#8203;:

> On Wed\, Apr 22\, 2015 4&#8203;:24&#8203;:52 PM\, public@&#8203;khwilliamson\.com wrote&#8203;:
> > On 04/20/2015 04&#8203;:12 PM\, Tony Cook via RT wrote&#8203;:
> > > On Mon Apr 20 10&#8203;:11&#8203;:55 2015\, public@&#8203;khwilliamson\.com wrote&#8203;:
> > >> On 04/20/2015 01&#8203;:21 AM\, Tony Cook \(via RT\) wrote&#8203;:
> > >>> The correct result \("a&#8203;:/"\) is produced if a non\-overloaded value is
> > >>> used\.
> > >>
> > >> Is this only on solaris?  Is it a 5\.22 blocker in your opinion?
> > >
> > > I think so\, both use64bitall and usethreads are common\.
> > >
> > > The smoke has been consistently failing since the commit\.
> > >
> > > I'll get you access to the VM where this is happening\, if you like\.
> > >
> > > Tony
> > 
> > The latest info about this is&#8203;:
> > 
> > Turning on compile and/or execute re debugging causes it to not fail\.
> > 
> > Compiling with \-O1 doesn't fail\.
> > 
> > perl won't fully compile with \-O0\, failing at the last moment in 
> > mktables \(while writing out the list of files that it changed\)\.
> > 
> > If I don't clean the workspace from an \-O1\, but re\-run Configure with 
> > instead \-O0\, then 'make test' \(so that mktables doesn't run\)\, the 
> > project compiles\, but with many many test failures\.
> 
> Just some additional information\,
> 
> TL;DR&#8203;: this bug does not seem to affect perl\-5\.21\.11 built with GCC\.
> 
> Yesterday I built 12 variants of perl 5\.21\.11 using GCC 4\.8\.4 and GCC
> 4\.9\.2\, on X86 32\-bit and 64\-bit Compile Servers\, running Sun Solaris
> 10 Update 8\, Oracle Solaris 10 Update 11 and Oracle Solaris 11\.2\. No
> failures seen in any "gmake test" runs\. These GCC were locally built
> not vendor provided\. The 32\-bit builds trigger a much larger set of
> warnings than the 64\-bit builds but no compilation aborts\.

Do you have the resources to run Test&#8203;::Smoke on these boxes? We do not
have that range of Solaris well\-covered\, so if that is an option\, it
would give us a nice addition to our test suite\.

> 64\-bit systems using \-DDuse64bitall
> 32\-bit systems using \-Duse64bitint
> All systems using \-Dusethreads \-Dusedtrace
> 
> Other options to change installation paths\, use bison\, and be explicit
> about a few Solaris default like \-Duselargefiles\.
> Before starting Configure\, removed all instances of \-G from
> hints/solaris\_2\.h \(overkill I know\) and used \-Dlddlflags='\-shared'
> 
> Regards\,
> Peter
> 
> PS&#8203;: VM Host names show the combinations explicitly\.
> 
> sol10u08\-gcc48\-x86\-32
> sol10u08\-gcc48\-x86\-64
> sol10u08\-gcc49\-x86\-32
> sol10u08\-gcc49\-x86\-64
> sol10u11\-gcc48\-x86\-32
> sol10u11\-gcc48\-x86\-64
> sol10u11\-gcc49\-x86\-32
> sol10u11\-gcc49\-x86\-64
> sol11u02\-gcc48\-x86\-32
> sol11u02\-gcc48\-x86\-64
> sol11u02\-gcc49\-x86\-32
> sol11u02\-gcc49\-x86\-64

\-\- 
H\.Merijn Brand  http&#8203;://tux\.nl   Perl Monger  http&#8203;://amsterdam\.pm\.org/
using perl5\.00307 \.\. 5\.21   porting perl5 on HP\-UX\, AIX\, and openSUSE
http&#8203;://mirrors\.develooper\.com/hpux/        http&#8203;://www\.test\-smoke\.org/
http&#8203;://qa\.perl\.org   http&#8203;://www\.goldmark\.org/jeff/stupid\-disclaimers/

**Comment by p5pRT** [comment]
### From pdb_ml@yahoo.com.au

On 24/04/15 06&#8203;:45 PM\, H\.Merijn Brand wrote&#8203;:
>
> Do you have the resources to run Test&#8203;::Smoke on these boxes? We do not
> have that range of Solaris well\-covered\, so if that is an option\, it
> would give us a nice addition to our test suite\.

Not at the moment\, the project I'm working on will keep these systems at 
very high utilization for the next few months\. I'll add a personal todo\, 
to look at the requirements after the project is complete\.

Regards\,
Peter

**Comment by p5pRT** [comment]
### From @rjbs

On Wed Apr 22 16&#8203;:24&#8203;:52 2015\, public@&#8203;khwilliamson\.com wrote&#8203;:
> The latest info about this is&#8203;:
> \[\.\.\.\]

So\, where do we stand?  The impression I get is that we don't have a clear route forward here\.  Any suggestions as to what can be done to move this forward?

\-\- 
rjbs

**Comment by p5pRT** [comment]
### From @tonycoz

On Mon Apr 27 15&#8203;:26&#8203;:16 2015\, rjbs wrote&#8203;:
> On Wed Apr 22 16&#8203;:24&#8203;:52 2015\, public@&#8203;khwilliamson\.com wrote&#8203;:
> > The latest info about this is&#8203;:
> > \[\.\.\.\]
> 
> So\, where do we stand?  The impression I get is that we don't have a
> clear route forward here\.  Any suggestions as to what can be done to
> move this forward?

This appears to be a code generation bug in solaris studio 12\.3\.

When calculating PL\_charclass\['\\n'\] it's using a bogus value from the stack
frame instead of looking up the value in the global PL\_charclass array\.

The non\-overloaded version is only working by accident\.

If I move the '\\n' to a variable the code works correctly\, of course\, this will slow down all implementations to work around what appears to be a real bug in
one compiler\.

An alternative might be to use C\< 0 > instead of C\< isWORDCHAR\_L1\('\\n'\) >
but I assume we're using the second for a reason \(it's self\-documenting
if nothing else\.\)

The code that's misbehaving is this code in S\_regmatch\(\)&#8203;:

&nbsp;                    case TRADITIONAL\_BOUND&#8203;:
&nbsp;                        ln = \(locinput == reginfo\->strbeg\)
&nbsp;                             ? isWORDCHAR\_L1\('\\n'\)
&nbsp;                             : isWORDCHAR\_utf8\(reghop3\(\(U8\*\)locinput\, \-1\,
&nbsp;                                                                \(U8\*\)\(reginfo\->strbeg\)\)\);
&nbsp;                        n = \(NEXTCHR\_IS\_EOS\)
&nbsp;                            ? isWORDCHAR\_L1\('\\n'\)
&nbsp;                            : isWORDCHAR\_utf8\(\(U8\*\)locinput\);
&nbsp;                        match = cBOOL\(ln \!= n\);
&nbsp;                        break;

The code generated for the isWORDCHAR\_L1\('\\n'\) appears to be&#8203;:

\{     5632  \}     movl    \-600\(%rbp\)\,%eax        / sym=\.CV12B
\{     5632  \}     andl    $1\,%eax        
\{     5632  \}     setne    %al        
\{     5632  \}     movsbl    %al\,%eax        / dest0=\#tmp496\, 

Tony

**Comment by p5pRT** [comment]
### From @khwilliamson

On 04/27/2015 05&#8203;:33 PM\, Tony Cook via RT wrote&#8203;:
> On Mon Apr 27 15&#8203;:26&#8203;:16 2015\, rjbs wrote&#8203;:
>> On Wed Apr 22 16&#8203;:24&#8203;:52 2015\, public@&#8203;khwilliamson\.com wrote&#8203;:
>>> The latest info about this is&#8203;:
>>> \[\.\.\.\]
>>
>> So\, where do we stand?  The impression I get is that we don't have a
>> clear route forward here\.  Any suggestions as to what can be done to
>> move this forward?
>
> This appears to be a code generation bug in solaris studio 12\.3\.
>
> When calculating PL\_charclass\['\\n'\] it's using a bogus value from the stack
> frame instead of looking up the value in the global PL\_charclass array\.
>
> The non\-overloaded version is only working by accident\.
>
> If I move the '\\n' to a variable the code works correctly\, of course\, this will slow down all implementations to work around what appears to be a real bug in
> one compiler\.
>
> An alternative might be to use C\< 0 > instead of C\< isWORDCHAR\_L1\('\\n'\) >
> but I assume we're using the second for a reason \(it's self\-documenting
> if nothing else\.\)

The only reason was to make it self\-documenting\, at no apparent runtime 
cost\, as the compiler should evaluate it and replace it by 0\.  But one 
can document it by doing what I do in the attached patch\, which I will 
smoke\.

I'm still concerned that compiling with \-O0 causes perl to not build\.  A 
final option is to use \-O1\, for the particular combination of Configure 
options that cause this failure\.  But what else in this compiler is broken?
>
> The code that's misbehaving is this code in S\_regmatch\(\)&#8203;:
>
>                      case TRADITIONAL\_BOUND&#8203;:
>                          ln = \(locinput == reginfo\->strbeg\)
>                               ? isWORDCHAR\_L1\('\\n'\)
>                               : isWORDCHAR\_utf8\(reghop3\(\(U8\*\)locinput\, \-1\,
>                                                                  \(U8\*\)\(reginfo\->strbeg\)\)\);
>                          n = \(NEXTCHR\_IS\_EOS\)
>                              ? isWORDCHAR\_L1\('\\n'\)
>                              : isWORDCHAR\_utf8\(\(U8\*\)locinput\);
>                          match = cBOOL\(ln \!= n\);
>                          break;
>
> The code generated for the isWORDCHAR\_L1\('\\n'\) appears to be&#8203;:
>
> \{     5632  \}     movl    \-600\(%rbp\)\,%eax        / sym=\.CV12B
> \{     5632  \}     andl    $1\,%eax        
> \{     5632  \}     setne    %al        
> \{     5632  \}     movsbl    %al\,%eax        / dest0=\#tmp496\,
>
> Tony
>
> \-\-\-
> via perlbug&#8203;:  queue&#8203;: perl5 status&#8203;: open
> https://rt-archive.perl.org/perl5/Ticket/Display.html?id=124348
>

**Comment by p5pRT** [comment]
### From @khwilliamson

<details><summary>0001-PATCH-perl-124348-XXX.patch</summary>

```diff
From 236603201bd10cef4fee81a23d74a1e9e7386316 Mon Sep 17 00:00:00 2001
From: Karl Williamson <khw@cpan.org>
Date: Mon, 27 Apr 2015 21:52:02 -0600
Subject: [PATCH] PATCH: [perl #124348] XXX

XXX Fill in
---
 regexec.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/regexec.c b/regexec.c
index 7247259..893ce3b 100644
--- a/regexec.c
+++ b/regexec.c
@@ -5623,11 +5623,11 @@ S_regmatch(pTHX_ regmatch_info *reginfo, char *startpos, regnode *prog)
                 switch((bound_type) FLAGS(scan)) {
                     case TRADITIONAL_BOUND:
                         ln = (locinput == reginfo->strbeg)
-                             ? isWORDCHAR_L1('\n')
+                             ? 0 /* isWORDCHAR_L1('\n') */
                              : isWORDCHAR_utf8(reghop3((U8*)locinput, -1,
                                                                 (U8*)(reginfo->strbeg)));
                         n = (NEXTCHR_IS_EOS)
-                            ? isWORDCHAR_L1('\n')
+                            ? 0 /* isWORDCHAR_L1('\n') */
                             : isWORDCHAR_utf8((U8*)locinput);
                         match = cBOOL(ln != n);
                         break;
@@ -5693,10 +5693,10 @@ S_regmatch(pTHX_ regmatch_info *reginfo, char *startpos, regnode *prog)
                 switch((bound_type) FLAGS(scan)) {
                     case TRADITIONAL_BOUND:
                         ln = (locinput == reginfo->strbeg)
-                            ? isWORDCHAR_L1('\n')
+                            ? 0 /* isWORDCHAR_L1('\n') */
                             : isWORDCHAR_L1(UCHARAT(locinput - 1));
                         n = (NEXTCHR_IS_EOS)
-                            ? isWORDCHAR_L1('\n')
+                            ? 0 /* isWORDCHAR_L1('\n') */
                             : isWORDCHAR_L1(nextchr);
                         match = cBOOL(ln != n);
                         break;
-- 
2.1.0


```
</details>

**Comment by p5pRT** [comment]
### From @khwilliamson

On 04/27/2015 09&#8203;:56 PM\, Karl Williamson wrote&#8203;:
> On 04/27/2015 05&#8203;:33 PM\, Tony Cook via RT wrote&#8203;:
>> On Mon Apr 27 15&#8203;:26&#8203;:16 2015\, rjbs wrote&#8203;:
>>> On Wed Apr 22 16&#8203;:24&#8203;:52 2015\, public@&#8203;khwilliamson\.com wrote&#8203;:
>>>> The latest info about this is&#8203;:
>>>> \[\.\.\.\]
>>>
>>> So\, where do we stand?  The impression I get is that we don't have a
>>> clear route forward here\.  Any suggestions as to what can be done to
>>> move this forward?
>>
>> This appears to be a code generation bug in solaris studio 12\.3\.
>>
>> When calculating PL\_charclass\['\\n'\] it's using a bogus value from the
>> stack
>> frame instead of looking up the value in the global PL\_charclass array\.
>>
>> The non\-overloaded version is only working by accident\.
>>
>> If I move the '\\n' to a variable the code works correctly\, of course\,
>> this will slow down all implementations to work around what appears to
>> be a real bug in
>> one compiler\.
>>
>> An alternative might be to use C\< 0 > instead of C\< isWORDCHAR\_L1\('\\n'\) >
>> but I assume we're using the second for a reason \(it's self\-documenting
>> if nothing else\.\)
>
> The only reason was to make it self\-documenting\, at no apparent runtime
> cost\, as the compiler should evaluate it and replace it by 0\.  But one
> can document it by doing what I do in the attached patch\, which I will
> smoke\.

I forgot to mention that doing this causes the the tests to not fail on 
Tony's machine that failed prior to the patch\.
>
> I'm still concerned that compiling with \-O0 causes perl to not build\.  A
> final option is to use \-O1\, for the particular combination of Configure
> options that cause this failure\.  But what else in this compiler is broken?
>>
>> The code that's misbehaving is this code in S\_regmatch\(\)&#8203;:
>>
>>                      case TRADITIONAL\_BOUND&#8203;:
>>                          ln = \(locinput == reginfo\->strbeg\)
>>                               ? isWORDCHAR\_L1\('\\n'\)
>>                               : isWORDCHAR\_utf8\(reghop3\(\(U8\*\)locinput\,
>> \-1\,
>>
>> \(U8\*\)\(reginfo\->strbeg\)\)\);
>>                          n = \(NEXTCHR\_IS\_EOS\)
>>                              ? isWORDCHAR\_L1\('\\n'\)
>>                              : isWORDCHAR\_utf8\(\(U8\*\)locinput\);
>>                          match = cBOOL\(ln \!= n\);
>>                          break;
>>
>> The code generated for the isWORDCHAR\_L1\('\\n'\) appears to be&#8203;:
>>
>> \{     5632  \}     movl    \-600\(%rbp\)\,%eax        / sym=\.CV12B
>> \{     5632  \}     andl    $1\,%eax
>> \{     5632  \}     setne    %al
>> \{     5632  \}     movsbl    %al\,%eax        / dest0=\#tmp496\,
>>
>> Tony
>>
>> \-\-\-
>> via perlbug&#8203;:  queue&#8203;: perl5 status&#8203;: open
>> https://rt-archive.perl.org/perl5/Ticket/Display.html?id=124348
>>
>

**Comment by p5pRT** [comment]
### From @khwilliamson

Thanks to Tony Cook for tracking this down

Fixed by not having the compiler do any evaluation
in commit 996de84d864e2452987ceb29dbf56f1ce92846b8

\-\- 
Karl Williamson

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'open' to 'pending release'

**Comment by p5pRT** [comment]
### From @khwilliamson

Thank you for submitting this ticket\.

The issue should now be resolved with the release today of Perl v5\.22\, which is available at http&#8203;://www\.perl\.org/get\.html
\-\- 
Karl Williamson for the Perl 5 team

**Comment by p5pRT** [comment]
@khwilliamson - Status changed from 'pending release' to 'resolved'
